### DISCUSSION TITLE
Stringify not working properly with data buckets

### DISCUSSION THREAD

**Discussion by marksyzm** [mockoon/mockoon#1487] [Help]
**Describe the bug**

When using the stringify function, it's not possible to import a data bucket and stringify the entire object data directly without stringify creating a rogue object with an "string" property that contains the stringified JSON.

**Steps to Reproduce**

Data bucket:
```
[1,2,3]
```

```
{ "data": {{{stringify (data 'qw6w')}}} }
```

**Should produce:**
```
{ "data": "[1,2,3]" }
```

**Actually produces:**
```
{ "data": { "string": "[1,2,3]" } }
```

**Mockoon version:**

8.4.0

**OS / OS version:**

Mac OS Sonoma 14.5 (23F79)

**Comment by 255kb** [comment]
`data` is already returning the data bucket content stringified. You either need to use `dataRaw`, or get rid of the `stringify` helper:

```handlebars
{ "data": {{data 'qw6w'}} }
```

**Comment by marksyzm** [comment]
Wrong, I require the data to be stringified as a property value. Note that in my example above, the first working example is stringified. 

For example, when I do this:
```json
{
  "someStringifiedJson": {{data 'gkcr'}}
}
```
This gives the undesired data array:
```json
{
  "someStringifiedJson": [1,2,3]
}
```
And not the desired:
```json
{
  "someStringifiedJson": "[1,2,3]"
}
```
Get it? Instead I am receiving:
```json
{
  "someStringifiedJson": {
    "string": "[1,2,3]"
  }
}
```

**Reply by 255kb** [reply]
Mockoon is built in JS and is using `JSON.stringify` for both the `data` and the `stringify` helper. 
When stringifying an object/array in JS, you will get a string containing the JSON representation of said object, but not surrounded by quotes.
i.e. you get `'[1,2,3]'` but not `'"[1,2,3]"'`

Also, if the result of the `data` helper goes through a `stringify` again it will give an unexpected result due to how Handlebars is working internally. 

If you want quotes you have to add them manually around your `data` helper: 

```
{
  "someStringifiedJson": "{{data 'gkcr'}}"
}
```

Both helpers are working as expected. 

--- 

As a side note: English is not my first language, but I find your tone a bit impatient, borderline aggressive. 
I will just say this: I have been working on Mockoon since 7 years. Maintaining and improving it during my free time, and now full-time since ~3.5 years. The app is used by thousands of people worldwide, hundreds of companies, and I must say that the amount of people actually contributing by opening PRs, trying to discuss improvements, or sponsoring my (full-time!) work on this tool is not null but close to 0. Acting like that won't make me fix bugs or add features faster. On the contrary.
PS: this applies to most open-source projects out there.

**Reply by marksyzm** [reply]
Apologies for my brashness; it was just that you just moved my bug into a discussion without looking at the actual bug, so it made me feel like my point was **_ignored_**, but actually perhaps it's just that I wasn't specific enough. I get that people work on things in their own spare time though.

My issue is that I *want* stringified data inside of JSON data, so not just arrays but any output from the inside of a data bucket. 

So for example, take this object in a data bucket:

```json
{
  "myObject": { "a": 1 }
}
```

I want to use this handlebars expression with that bucket:

```json
{
  "stringifiedJsonData": {{{stringify (data 'abcd')}}}
}
```
To create this desired output:
```json
{
  "stringifiedJsonData": "{\n  \"myObject\": { \"a\": 1  }\n}"
}
```

Instead I get this, which I see as a bug:

```json
{
  "stringifiedJsonData": {
    "string": "{\n  \"myObject\": { \"a\": 1  }\n}"
  }
}
```

**Reply by marksyzm** [reply]
I'll do anything I can to contribute :) I can see this is a good tool and I do contribute to other repos.

**Reply by 255kb** [reply]
Apologies accepted :) 
The initial description of the issue let me think that everything was working as intended. I prefer when people open discussions first so we can decide if there is a bug or something to improve. It's easier for me to triage.

If you want to add escaped JSON inside a string inside JSON, I think that stringifying twice (`stringify` + `data`) is the right approach. 
However, to avoid Handlebars escaping and having to use triple slashes, the `data` helper is returning a SafeString. But the `stringify` helper is not playing well with a SafeString, thus the `{ "string": ...}` format. 

I think updating the `stringify` helper to handle correctly SafeString could do the trick. I tried locally to change the `stringify` helper behavior, with the following data bucket: 

```json
{
  "myarr": [1,2,3]
}
```

And the following response body: 

```handlebars
{
  "result": {{{stringify (data 'data')}}}
}
```

And It gave me: 

```json 
{
  "result": "{\"myarr\":[1,2,3]}"
}
```

I think this looks like it would work?

I will open an issue. It's a very tiny fix. However, the next release won't be before end of September due to a number of factors. You will have to build the app from the sources if you want to use the fix.

**Reply by marksyzm** [reply]
Perfect, thank you so much! This will be just in time for our MVP :) I'll check with the team about the version and patching. Thanks for your patience with me!

**Reply by 255kb** [reply]
Fix was merged if you want to build from the sources: https://github.com/mockoon/mockoon/pull/1490

**Reply by marksyzm** [reply]
Thanks so much for your hard work

**Reply by marksyzm** [reply]
I'll make a donation as soon as i get paid this week 😀

**Reply by 255kb** [reply]
Thank you for your support! It means a lot 🤗

**Comment by marksyzm** [comment]
I am reporting this as a bug in mockoon where it won't stringify a data bucket to a value.

To be absolutely specific, my case here is that I need the data from another data bucket to be stringified in order to emulate what I use for a client's SAP API which returns the JSON data stringified in another JSON object. 

In your case, it does this to some extent but returns it as an object with a property `"string"` and the stringified JSON when it should just return a string.

**Reply by marksyzm** [reply]
@255kb I'm going to give this another day and then I'll create this as an issue again as I need this specific functionality.
